### ISSUE TITLE
Patch does not update the data

### ISSUE THREAD

**Issue by incomudro** [json-api-dotnet/JsonApiDotNetCore#514]
## Patch does not update the data

We are building an API in AspNetCore (without a mapping for the entities). While using the `JsonApiController`, the Get, Post en Delete work fine. However, the Patch does not work. 

I have override the PatchAsync, and the application get into that breakpoint, but the entity remains empty and the `AttributesToUpdate` remains an empty list.

I seems as if this code should work (even without the overriding method in the controller). 
What am I missing, or what could I test?

------ _see (parts of) the code below._ ------

| Key | Value | 
| --------------  |:-------------:|
| Type               | PATCH |
| URL                | http://localhost:62596/patchtable/2|
| Content-Type | application/vnd.api+json |
| Body | {"data": {"type": "patchtable","id": 2,"attributes": {"text": "A new value"}}} |

The code inside the `PatchtableController.cs`. In comments I've added some of the findings I've made when trying to debug what was going on.
```
    public class PatchtableController : JsonApiController<Patchtable>
    {
        private IJsonApiContext _jsonApiContext;

        public PatchtableController(
            IJsonApiContext jsonApiContext, 
            IResourceService<Patchtable, int> resourceService) :
            base(jsonApiContext, resourceService)
        {
            _jsonApiContext = jsonApiContext;
        }

        [HttpPatch("{id}")]
        public override async Task<IActionResult> PatchAsync(int id, [FromBody] Patchtable entity)
        {
            //  _jsonApiContext.AttributesToUpdate.Count -> 0;
            //  Patchtable entity -> null

            return await base.PatchAsync(id, entity);
        }
    }
```

The test-model is just a simple without any relations. See the `Patchtable .cs`, the `PatchtableTypeConfiguration ` below and part of our `ApplicationDbContext.cs` below.
```
    public class Patchtable : Identifiable
    {
        [Attr("text")]
        [Required]
        public string Text { get; set; }
    }
```
```
    public class PatchtableTypeConfiguration : IEntityTypeConfiguration<Patchtable>
    {
        public void Configure(EntityTypeBuilder<Patchtable> builder)
        {
            builder.HasKey(x => x.Id);

            builder.Property(x => x.Text)
                .IsRequired();
        }
    }
```
```
    public class ApplicationDbContext : DbContext
    {
        public ApplicationDbContext(DbContextOptions options) : base(options) { }

        public DbSet<...> ...{ get; set; }
        public DbSet<Patchtable> Patchtables { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.ApplyConfiguration(new ...TypeConfiguration());
            modelBuilder.ApplyConfiguration(new PatchtableTypeConfiguration());
        }
    }
```

See the `Startup.cs` below. I've tested the application with adding the option `options.EnableOperations = true`, without result.
```
    public void ConfigureServices(IServiceCollection services)
    {
        string connection = Configuration["ConnectionStrings:DefaultConnection"];
        string assemblyName = Assembly.GetAssembly(typeof(ApplicationDbContext)).FullName;
        services.AddDbContext<ApplicationDbContext>(options =>
            options.UseSqlServer(
                connection,
                sqlServerOption => sqlServerOption.MigrationsAssembly(assemblyName)));
        services.AddJsonApi<ApplicationDbContext>();
        services.AddOpenApiDocument(settings =>
            settings.SchemaType = SchemaType.OpenApi3);
        services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
    }
```

...

## Environment

- JsonApiDotNetCore version: 3.1.0
- Microsoft.AspNetCore.App version: 2.2.0
- Microsoft.EntityFrameworkCore.SqlServer version: 2.2.4
- Microsoft.NETCore.App version: 2.2.0

**Comment by wisepotato** [comment]
I dont personally use the entity resource split, maybe @milosloub has some insights into this?

**Comment by incomudro** [comment]
> 
> 
> I dont personally use the entity resource split, maybe @milosloub has some insights into this?

Entity resource split? In this example the model in the API is equal to the data-model and am not aware of any splitting. But it can be that I misunderstand with is meant with 'entity split'. 

For testing the patch operation loose from the rest of the application, I created a simple model and called it `patchtable`, that would be stored in he database table named `patchtable`. With hindsight I suspect that another name for the test-model might have been better.

**Comment by maurei** [comment]
Weird, I can't see anything odd in your code. Could you share your patchable example app in a repo so I can run it and dive into the details?

**Comment by incomudro** [comment]
@maurei  I've created an empty project with just one entity. The project is public, thus you should be able to view its content. EF is setup with a MS SQL Server in mind.  If preferred, I can use the source-code of your application and add it to the project, instead of adding it as a NuGet package. In the readme I've included the API calls I've tried. 

I suspect that I am missing something somewhere, but I seem to fail to see what it is.

https://github.com/incomudro/ApiTestProject

**Comment by maurei** [comment]
I'll look into it and let you know!

**Comment by milosloub** [comment]
@incomudro Hi, be carefull with your naming convention - your Context use Patchables and your type is "patchable". The problem with this line
`services.AddJsonApi<ApplicationDbContext>();`
is, that your application muse use same convention in controller layer as you use in your ApplicationDbContext.
And there is another hidden convention with this - your controller name defines "type"
(BlogController => type="blog", BlogsController => type="blogs")

First approach to fix- pluralize
1) Change your PatchtableController -> PatchtablesController 
2) Change your url to: http://localhost:62596/patchtables/2
3) Change your type in body from "patchable" to "patchables"

Second approach to fix - singularize
1) Change your DbSet<Patchtable> Patchtables -> DbSet<Patchtable> Patchtable

**Comment by incomudro** [comment]
Thanks for the feedback, although the answer contained a mistake (see second lesson learned). Basically, the type in the body of the patch-request should match the name of the associated DbSet<T>, as in, both should be either plural or both should be singular. 

The feedback did however helped me in the right direction :)

I've updated my test-project and will leave it there for a while. Feel free to pick anything from it should you wish to add another example to your list of example-projects. In the readme I've added two lessons-learned, the things that went wrong while trying to figure out what went wrong.
https://github.com/incomudro/ApiTestProject

**Lesson 1: Case-sensitivity**
For PATCH, the attribute is case-sensitive and should be the same as defined in the model. As the attribute is not case-sensitive for POST, this mistake was probably overlooked once or twice.

**Lesson 2: Type**
The type defined in the PATCH, should be the same as the name for the `DbSet<T>`. Thus when the `DbSet<Fruit>` is 'Fruits', the type provided in the body of the PATCH should be 'fruits' as well. Know that the type is not case-sensitive and both `"type": "fruits"` and `"type": "Fruits"` will work.
